src/wire/icmpv6.rs

```diff
@@ -0,0 +1,690 @@
+    pub doc enum DstUnreachable(u8) {
+        /// No Route to destination.
+        NoRoute         = 0,
+        /// Communication with destination administratively prohibited.
+        AdminProhibit   = 1,
+        /// Beyond scope of source address.
+        BeyondScope     = 2,
+        /// Address unreachable.
+        AddrUnreachable = 3,
+        /// Port unreachable.
+        PortUnreachable = 4,
+        /// Source address failed ingress/egress policy.
+        FailedPolicy    = 5,
+        /// Reject route to destination.
+        RejectRoute     = 6
+    }
+}
+
+
+impl fmt::Display for DstUnreachable {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &DstUnreachable::NoRoute =>
+                write!(f, "no route to destination"),
+            &DstUnreachable::AdminProhibit =>
+                write!(f, "communication with destination administratively prohibited"),
+            &DstUnreachable::BeyondScope =>
+                write!(f, "beyond scope of source address"),
+            &DstUnreachable::AddrUnreachable =>
+                write!(f, "address unreachable"),
+            &DstUnreachable::PortUnreachable =>
+                write!(f, "port unreachable"),
+            &DstUnreachable::FailedPolicy =>
+                write!(f, "source address failed ingress/egress policy"),
+            &DstUnreachable::RejectRoute =>
+                write!(f, "reject route to destination"),
+            &DstUnreachable::Unknown(id) =>
+                write!(f, "{}", id)
+        }
+    }
+}
+
+/// A read/write wrapper around an Internet Control Message Protocol version 6 packet buffer.
+#[derive(Debug, PartialEq)]
+pub struct Packet<T: AsRef<[u8]>> {
+    buffer: T
+}
+
+// Ranges and constants describing key boundaries in the ICMPv6 header.
+// See https://tools.ietf.org/html/rfc4443 for details.
+mod field {
+    use wire::field::*;
+
+    pub const TYPE:       usize = 0;
+    pub const CODE:       usize = 1;
+    pub const CHECKSUM:   Field = 2..4;
+
+    pub const UNUSED:     Field = 4..8;
+    pub const MTU:        Field = 4..8;
+    pub const POINTER:    Field = 4..8;
+    pub const ECHO_IDENT: Field = 4..6;
+    pub const ECHO_SEQNO: Field = 6..8;
+}
+
+impl<T: AsRef<[u8]>> Packet<T> {
+    /// Imbue a raw octet buffer with ICMPv6 packet structure.
+    pub fn new(buffer: T) -> Packet<T> {
+        Packet { buffer }
+    }
+
+    /// Shorthand for a combination of [new] and [check_len].
+    ///
+    /// [new]: #method.new
+    /// [check_len]: #method.check_len
+    pub fn new_checked(buffer: T) -> Result<Packet<T>> {
+        let packet = Self::new(buffer);
+        packet.check_len()?;
+        Ok(packet)
+    }
+
+    /// Ensure that no accessor method will panic if called.
+    /// Returns `Err(Error::Truncated)` if the buffer is too short.
+    pub fn check_len(&self) -> Result<()> {
+        let len = self.buffer.as_ref().len();
+        if len < self.header_len() {
```

Comment:
I'm not sure if this is correct. `header_len` returns results more similar to the older version of ICMPv4's `header_len` (can be 4 or 8). This is so that `payload` can be more useful for NDISC and MLDv2 etc that need the last 4 bytes of the header. That being said, no valid ICMPv6 message is less than 8 bytes in length.

Reply:
This will actually panic if you supply an empty buffer because `header_len` checks `msg_type`. If no valid ICMPv6 message is less than 8 bytes in length you should just use `field::UNUSED.end` (actually I prefer it to be called `field::HEADER_END`.

Reply:
:+1: